File: compiler/nodes/types/parameter.go
package types

import (
	"wipple/database"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type TypeParameterNode struct {
	Name  string
	Infer bool
	Value database.Node
	Facts *database.Facts
}

func (node *TypeParameterNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseTypeParameters(parser *syntax.Parser) ([]TypeParameterNode, *syntax.Error) {
	many, err := syntax.ParseMany(parser, 0, ParseTypeParameter, syntax.ParseNothing)
	if err != nil {
		return nil, err
	}

	if len(many) > 0 {
		_, err := parser.Token("TypeFunctionOperator", syntax.TokenConfig{
			Commit: "in this generic item",
		})

		if err != nil {
			return nil, err
		}

		parser.ConsumeLineBreaks()
	}

	parameters := make([]TypeParameterNode, 0, len(many))
	for _, item := range many {
		parameters = append(parameters, *item.Value)
	}

	return parameters, nil
}

func ParseTypeParameter(parser *syntax.Parser) (*TypeParameterNode, *syntax.Error) {
	return syntax.ParseCached(parser, func(p *syntax.Parser) (*TypeParameterNode, *syntax.Error) {
		parameter, ok, err := syntax.ParseOptional(parser, ParseNamedTypeParameter)
		if err != nil {
			return nil, err
		}
		if ok {
			return parameter, nil
		}

		parameter, ok, err = syntax.ParseOptional(parser, ParseInferTypeParameter)
		if err != nil {
			return nil, err
		}
		if ok {
			return parameter, nil
		}

		return nil, parser.Error("Expected type parameter")
	})
}

func ParseAnnotatedParameterType(parser *syntax.Parser) (*TypeParameterNode, *syntax.Error) {
	span := parser.Spanned()

	name, err := syntax.ParseTypeParameterName(parser)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("AnnotateOperator", syntax.TokenConfig{
		Commit: "in this type annotation",
	})
	if err != nil {
		return nil, err
	}

	parser.ConsumeLineBreaks()

	value, err := ParseType(parser)
	if err != nil {
		return nil, err
	}

	return &TypeParameterNode{
		Name:  name,
		Infer: false,
		Value: value,
		Facts: database.NewFacts(span()),
	}, nil
}

func ParseParameterType(parser *syntax.Parser) (*TypeParameterNode, *syntax.Error) {
	span := parser.Spanned()

	name, err := syntax.ParseTypeParameterName(parser)
	if err != nil {
		return nil, err
	}

	return &TypeParameterNode{
		Name:  name,
		Infer: false,
		Value: nil,
		Facts: database.NewFacts(span()),
	}, nil
}

func ParseNamedTypeParameter(parser *syntax.Parser) (*TypeParameterNode, *syntax.Error) {
	span := parser.Spanned()

	name, err := syntax.ParseTypeParameterName(parser)
	if err != nil {
		return nil, err
	}

	return &TypeParameterNode{
		Name:  name,
		Infer: false,
		Value: nil,
		Facts: database.NewFacts(span()),
	}, nil
}

func ParseInferTypeParameter(parser *syntax.Parser) (*TypeParameterNode, *syntax.Error) {
	span := parser.Spanned()

	_, err := parser.Token("LeftParenthesis", syntax.TokenConfig{
		Reason: "between these parentheses",
	})
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("InferKeyword", syntax.TokenConfig{
		Commit: "in this inferred type parameter",
	})
	if err != nil {
		return nil, err
	}

	name, err := syntax.ParseTypeParameterName(parser)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("RightParenthesis")
	if err != nil {
		return nil, err
	}

	return &TypeParameterNode{
		Name:  name,
		Infer: true,
		Value: nil,
		Facts: database.NewFacts(span()),
	}, nil
}

func (node *TypeParameterNode) Visit(visitor *visit.Visitor) {
	visitType(visitor, node)

	existing, ok := visit.Resolve[*visit.TypeParameterDefinition](visitor, node.Name, node)
	if ok {
		visitor.Constraint(typecheck.NewGroupConstraint(node, existing.Node))
	} else {
		definition := &visit.TypeParameterDefinition{
			Name: node.Name,
			Node: node,
		}
		visitor.Define(node.Name, definition)

		visitor.Constraint(typecheck.NewTypeConstraint(node, typecheck.ParameterType(node, node.Name)))

		if node.Infer {
			database.SetFact(node, typecheck.InferredParameterFact{})
		}

		// Update the `Resolved` fact
		visit.Resolve[*visit.TypeParameterDefinition](visitor, node.Name, node)

		if node.Value != nil {
			visitor.Visit(node.Value)

			constraint := typecheck.NewGroupConstraint(node, node.Value)
			constraint.Info().IsActive = false // wait until instantiated
			visitor.Constraint(constraint)
		}

		typeParameters, _ := database.GetFact[visit.TypeParametersFact](visitor.CurrentDefinition.Node)
		typeParameters = append(typeParameters, node)
		database.SetFact(visitor.CurrentDefinition.Node, typeParameters)
	}
}

File: compiler/src/nodes/types/parameter.rs
use crate::{
    codegen::Codegen,
    database::{Node, NodeRef},
    nodes::{parse_type, visit_type},
    syntax::{ParseError, Parser, TokenKind, parse_type_parameter_name},
    typecheck::{GroupConstraint, InferredParameter, TypeConstraint},
    visit::{Definition, TypeParameterDefinition, TypeParameters, Visit, Visitor},
};

#[derive(Debug)]
pub struct TypeParameterNode {
    pub name: String,
    pub infer: bool,
    pub value: Option<NodeRef>,
}

impl Node for TypeParameterNode {}

pub fn parse_type_parameter(parser: &mut Parser<'_>) -> Result<TypeParameterNode, ParseError> {
    if let Some(node) = parser.parse_optional(parse_named_type_parameter)? {
        return Ok(node);
    }

    if let Some(node) = parser.parse_optional(parse_infer_type_parameter)? {
        return Ok(node);
    }

    Err(parser.error("Expected type parameter"))
}

pub fn parse_annotated_type_parameter(
    parser: &mut Parser<'_>,
) -> Result<TypeParameterNode, ParseError> {
    let name = parse_type_parameter_name(parser)?;
    parser.token(TokenKind::AnnotateOperator)?;
    parser.commit("in this type annotation");
    parser.consume_line_breaks();
    let value = parse_type(parser)?;

    Ok(TypeParameterNode {
        name,
        infer: false,
        value: Some(value),
    })
}

pub fn parse_annotated_type_parameter_node(parser: &mut Parser<'_>) -> Result<NodeRef, ParseError> {
    let span = parser.spanned();
    let node = parse_annotated_type_parameter(parser)?;
    let span = span(parser);
    Ok(parser.register(span, node))
}

pub fn parse_named_type_parameter(
    parser: &mut Parser<'_>,
) -> Result<TypeParameterNode, ParseError> {
    let name = parse_type_parameter_name(parser)?;

    Ok(TypeParameterNode {
        name,
        infer: false,
        value: None,
    })
}

pub fn parse_infer_type_parameter(
    parser: &mut Parser<'_>,
) -> Result<TypeParameterNode, ParseError> {
    parser.token_with_reason(TokenKind::LeftParenthesis, "between these parentheses")?;
    parser.token(TokenKind::InferKeyword)?;
    let name = parse_type_parameter_name(parser)?;
    parser.token(TokenKind::RightParenthesis)?;

    Ok(TypeParameterNode {
        name,
        infer: true,
        value: None,
    })
}

pub fn parse_type_parameters(parser: &mut Parser<'_>) -> Result<Vec<NodeRef>, ParseError> {
    let parameters = parser
        .parse_many(
            0,
            |parser| parser.node(parse_type_parameter),
            |parser| parser.parse_nothing(),
        )?
        .into_iter()
        .map(|(node, _)| node)
        .collect::<Vec<_>>();

    if !parameters.is_empty() {
        parser.token(TokenKind::TypeFunctionOperator)?;
        parser.consume_line_breaks();
    }

    Ok(parameters)
}

impl Visit for TypeParameterNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_type(node, visitor);

        let existing = visitor.resolve(&self.name, node, |definition| match definition {
            Definition::TypeParameter(definition) => Some(definition.clone()),
            _ => None,
        });

        if let Some(existing) = existing {
            visitor.constraint(GroupConstraint::new(node.clone(), existing.node));
        } else {
            let definition = TypeParameterDefinition {
                name: self.name.clone(),
                node: node.clone(),
            };

            visitor.define(&definition.name, definition.clone());

            if let Some(value) = &self.value {
                visitor.visit(value);
                visitor.edge(value, node, "value");

                let mut constraint = GroupConstraint::new(node.clone(), value.clone());
                constraint.info_mut().active = false; // wait until instantiated
                visitor.constraint(constraint);
            }

            if self
                .value
                .as_ref()
                .and_then(|value| value.downcast_ref::<TypeParameterNode>())
                .is_some()
            {
                // Allow aliases to type parameters
            } else {
                visitor.constraint(TypeConstraint::new(
                    node.clone(),
                    visitor.parameter_type(node.clone()),
                ));
            }

            if self.infer {
                visitor.insert(node, InferredParameter);
            }

            // Update the `Resolved` fact
            visitor.resolve(&self.name, node, |definition| match definition {
                Definition::TypeParameter(_) => Some(()),
                _ => None,
            });

            let definition_node = visitor.current_definition().node.clone();

            visitor.with_fact(&definition_node, |TypeParameters(parameters)| {
                parameters.push(node.clone())
            });
        }
    }
}

impl Codegen for TypeParameterNode {}
